docs/contributing/contributing-conventions.rst

```diff
@@ -0,0 +1,128 @@
+
+    - in a function, the top-most scope is commonly used for
+      argument-unpacking, and the usual ``int err`` variable, along with
+      additional variable needed for the function to do its thing
+    - using variable initialization in for-loops is perfectly fine
+    - variables in nested-block-scope is fine; still at the top of that
+      block-scope, e.g. the top of a loop-nest
+
+* Variable initialization
+
+  - Do not initialize variables using functions that require error-handling
+  - Limit to argument unpacking
+  - Limit to static assignment
+
+* Global variables
+
+  - Prefix them with ``g_``
+
+C: API
+------
+
+* The API and its backend implementations aims to be re-entrant
+  - Pass pointers to "objects" e.g. ``xnvme_cmd_pass(*ctx, ...);``
+  - Pass double-pointers for "object-initialization" e.g. ``xnvme_queue(..., **queue);``
+  - On success, return 0. On error, Return negative ``errno``.
+  - Exceptions: API-calls mimicking legacy interfaces, such as the
+    ``xnvme_buf_alloc()`` which is operating similar to ``malloc()/free()``.
+
+* The API and its backend implementation cannot be assumed to be thread-safe
+
+* Be minimal with definitions in the public API
+
+  - E.g. the 'sys/queue.h' was removed from the public API as it gave several
+    head-aches when building xNVMe on multiple platforms. However, internally,
+    specifically in backend implementations, assumptions can be made on the
+    availability of certain headers and their general availability
+
+* Each function in the API must have a command-line tool exercising it
+
+  - The source-code for the command-line tools serve as examples of utilizing
+    the C API in general
+  - This is essential for the NVMe command-set APIs as a means of demonstrating
+  - The command-line utilities also serve as a convenient means to interact
+    with NVMe devices via the command-line
+  - The command-line utilities must use the ``libxnvmec.h`` as this provides a
+    common command-line interface, the common-cli is nice as for usability,
+    such that all cli-tools use the same arguments, it is also essential for
+    instrumentation during testing
+
+C: Backend Implementations
+--------------------------
+
+The backend interface is declared in the internal header ``xnvme_be.h``. It
+consists of function-pointers for different tasks grouped together in structs.
+Those different set of tasks are referred to as the function-interfaces, they
+are labeled / group by:
+
+* **dev** : device handles and device enumeration
+* **mem** : memory allocation for command-payloads aka buffers
+* **admin** : synchronous admin-command processing
+* **sync** : synchronous I/O command processing
+* **async** : asynchronous I/O command submission/completion via queues
+
+Additionally, a backend-state is available for specialization to a given
+backend and its function-interface implementations.
+
+* Have a header-file defining the backend and what it exposes
```

Comment:
I think this should be prefixed with "Should" to follow the convention of the following sections.

`* Should have a header-file defining the backend and what it exposes`

Reply:
Changed, thanks!